Added Grafana-agent page for the decision on those

MadsHT opened the pull request:
Added my verdict on the Grafana agent.

Will update decision log when decision is made.

klausenbusk commented on docs/applications/grafana-agent.md:
```diff
@@ -0,0 +1,22 @@
+- Documentation is confusing
```
Perhaps we should put the file under `docs/decisions/0001-grafana-agents.md` and create some sort of "template" for decisions (summary, rational etc.) (not sure).

klausenbusk commented on docs/applications/grafana-agent.md:
```diff
@@ -0,0 +1,22 @@
+- We send all logs to one place
+- We want all metrics to be scraped by one agent anyways.
+- Curently configuration options are lackluster
+
```
Perhaps we should mention why we chose grafana agents in the first place.  IIRC one of the reason was we could avoid a big Prometheus instance, when we basically just needs `remote write`. That is something Prometheus supports now (https://prometheus.io/blog/2021/11/16/agent/).

klausenbusk commented on docs/applications/grafana-agent.md:
```diff
@@ -0,0 +1,22 @@
+
+<br>
+
+Why should we change back to prometheus and promtail
```
https://prometheus.io/blog/2021/11/16/agent/ is worth mentioning here.

skovbakke commented on docs/decisions/0001-grafana-agents.md:
```diff
@@ -0,0 +1,26 @@
+
+The reason we choose grafana agents is that we wanted a lightweight way of using the `remote write` feature to write to cortex, also this would allow us to get rid of a bulky Prometheus instance.
+
+<br>
```
Does this actually do anything in markdown?

skovbakke commented:
In regards to having a template. Kristian suggested, that we create a template for decisions. I think this is a very good idea, perhaps we should create it at a workshop at some point.

MadsHT commented:
According to this GH issue [Prometheus-operator 3989](https://github.com/prometheus-operator/prometheus-operator/issues/3989) The promethus operator can modified to run in agent mode, but it requires manuel deletion of parts of the config.

From what I understand there will start work on a PrometheusAgent CRD statet [here](https://github.com/prometheus-operator/prometheus-operator/issues/3989#issuecomment-1015187456)
Though [in this comment](https://github.com/prometheus-operator/prometheus-operator/issues/3989#issuecomment-1018374466) they will not commit to a ETA.

So I propose that we for now spin up a normal prometheus instance and keep an eye on the Prometheus Agent CRD Issue

klausenbusk commented:
> According to this GH issue [Prometheus-operator 3989](https://github.com/prometheus-operator/prometheus-operator/issues/3989) The promethus operator can modified to run in agent mode, but it requires manuel deletion of parts of the config.

AFAIU that was implemented in https://github.com/prometheus-operator/prometheus-operator/pull/4417, and we just need to override the args and it should work™ (example here: https://github.com/prometheus-operator/kube-prometheus/blob/f6e704cceb09de9c14db3d5bcb6fb1ea1217a040/examples/prometheus-agent.jsonnet#L24-L34).

MadsHT commented:
> The blocking issue right now is that the generated configuration isn't compatible with the agent mode because the operator generates the rules and alerting sections which isn't permitted for the agent. As a workaround, I've set the resource in paused mode, modified the generated secret manually to remove the 2 keys and it worked!
I think that it would be fine to tweak the generated Prometheus configuration so that rules and alerting fields aren't present when the custom resource has nil rule selectors and alerting spec.

This sentance states that the generated config contains fields that is not permitted in the agent.
Stating that: 
>As a workaround, I've set the resource in paused mode, modified the generated secret manually to remove the 2 keys and it worked!

And I feel like that is not what we're looking for right now?

klausenbusk commented:
> And I feel like that is not what we're looking for right now?

The PR (https://github.com/prometheus-operator/prometheus-operator/pull/4417) was created after that comment, so I don't think the comment is accurate anymore. If you look at the PR you can also see it removes all the "problematic" sections. The agent example was also added after that PR was merged: https://github.com/prometheus-operator/kube-prometheus/pull/1472.

MadsHT commented:
Well you would have to disabel the `ruleSelector` and `alerting` object, and atleast the `alerting` obj is required by the helm CRD

![image](https://user-images.githubusercontent.com/31923273/156352771-33c38226-bc4d-4845-8b22-c07a23505bd2.png)

MadsHT commented:
In this issur [[prometheus-kube-stack] Prometheus in Agent Mode](https://github.com/prometheus-community/helm-charts/issues/1519) [This comment](https://github.com/prometheus-community/helm-charts/issues/1519#issuecomment-1017241942) states that the helm chart will wait for the final solution, i.e. the prometheusAgent CRD before doing anything with this?

klausenbusk commented:
Hmm, if it is just fixing the chart, we should consider sending a PR, but it can wait. Let's get it working first :)

MadsHT commented:
I think the fix is simple, removing these lines
https://github.com/prometheus-community/helm-charts/blob/90930fcb26e41156892a4457133b38a238469ee8/charts/kube-prometheus-stack/crds/crd-prometheuses.yaml#L1172-L1174

Other then that I have prometheus running.

klausenbusk commented on docs/decisions/0001-grafana-agents.md:
```diff
@@ -0,0 +1,26 @@
+- Documentation is confusing
```
```suggestion
# Grafana Agents - go/no go?

Reasons for ditching Grafana Agents:
- Documentation is confusing
```

klausenbusk commented on docs/decisions/0001-grafana-agents.md:
```diff
@@ -0,0 +1,26 @@
+<br>
+
+Why should we change back to prometheus and promtail
+- Prometheus has an agent feature in experimental, and according to [this blog](https://prometheus.io/blog/2021/11/16/agent/) it has been tested at scale.
```
```suggestion
- Prometheus has an experimental agent feature, and according to [this blog](https://prometheus.io/blog/2021/11/16/agent/) it has been tested at scale
```

MadsHT commented on docs/decisions/0001-grafana-agents.md:
```diff
@@ -0,0 +1,26 @@
+- Documentation is confusing
+- Agent doc and operator doc is not aligned
+- Doc makes working with this developer hostile
```
The developers in our team?

What I was trying to say is that working with documentation that are not aligend and more confusing then any thing else is developer hostile, from my point of view?

klausenbusk replied:
Hmm, I see. The sentence isn't easy to read (IMO).

Could we just change:
```
- Documentation is confusing
- Agent doc and operator doc is not aligned
- Doc makes working with this developer hostile
```

to

```
- The documentation is confusing and the Agents and operator docs aren't aligned
```

MadsHT replied:
Works for me :+1 and sounds better

MadsHT commented on docs/decisions/0001-grafana-agents.md:
```diff
@@ -0,0 +1,26 @@
+- Functionality we though we wanted to use is not pressent - integrations and traces
+- In the current Agent configuration, we're using it as a package deal for promtail and prometheus.
+- We send all logs to one place
+- We want all metrics to be scraped by one agent anyways.
```
That the proposed solution that was build did not utilize the part where we could spin up Grafana agents anywhere, that setup just had one agent doing all the scraping

klausenbusk replied:
Per cluster or per node?

MadsHT replied:
The metrics scraping was done with one pr. cluster, the log scraping was one pr. node

klausenbusk replied:
Hmm, okay. I think this should be reworded and/or moved somewhere else in the document. It sounds more like a fact than a argument against Grafana Agents.
```
- In the current Agent configuration, we're using it as a package deal for promtail and prometheus.
- We send all logs to one place
- We want all metrics to be scraped by one agent anyways.
```

klausenbusk commented on docs/decisions/0001-grafana-agents.md:
```diff
@@ -0,0 +1,27 @@
+- In the current Agent configuration, we're using it as a package deal for promtail and prometheus.
+- We send all logs to one place
+- We want all metrics to be scraped by one agent anyways.
+- Curently configuration options are lackluster
```
Anything is particular missing?

klausenbusk commented on docs/decisions/0001-grafana-agents.md:
```diff
@@ -0,0 +1,27 @@
+- The prometheus operator could also be worth looking at since it give us some of the same features as the agents does, while still providing higher configuration options.
+  (Also endorsment for prometheus-operator is higher - 6,7k stars vs 594 at the time of writing)
+- For Promtail, I am currently seeing issues related to `filetarget.fsnotify.newwatcher: to many open files` which is said to be fixed by using `fs.inotify.max_user_instances` but the Grafana agents does not have a way of adding this to the logging pods, this promtail does [here](https://github.com/grafana/helm-charts/blob/e64169fab5fe248787bd62f4e1a990760fbf99a8/charts/promtail/values.yaml#L20)
+
+So from my point of view we should consider going back and use prometheus and promtail,
+
+This is manily due to the inconsistent doc and early stage of the grafana agents, and we need something that is more mature and has more consisten doc.
+
+I would suggest using the prometheus operator for convinience and then deploying promtail seperate to Loki.
```
```suggestion

### Recommendation

So from my point of view we should consider going back and use Prometheus and promtail,

This is mainly due to the inconsistent doc and early stage of the Grafana Agents, and we need something that is more mature and has more consistent doc.

I would suggest using the prometheus operator for convenience and then deploying promtail separate to Loki.
```

MadsHT replied:
For the flexibility, I suppose either way if we want to use the distributed version of Loki, then we would have to build a dependency into our proxy chart.

But mostly the flexibility?

klausenbusk commented on docs/decisions/0001-grafana-agents.md:
```diff
@@ -0,0 +1,32 @@
+
+This is manily due to the inconsistent doc and early stage of the grafana agents, and we need something that is more mature and has more consisten doc.
+
+I would suggest using the prometheus operator for convinience and then deploying promtail seperate to Loki.
```
```suggestion
I would suggest using the prometheus operator for convinience and then deploying promtail separate to Loki for extra flexibility, so we don't need to touch Loki if we want to switch the logging agent (ex: to Grafana Agents when it is more mature).
```

klausenbusk commented on docs/decisions/0001-grafana-agents.md:
```diff
@@ -0,0 +1,32 @@
+
+This is manily due to the inconsistent doc and early stage of the grafana agents, and we need something that is more mature and has more consisten doc.
+
+I would suggest using the prometheus operator for convinience and then deploying promtail separate to Loki for extra flexibility, so we don't need to touch Loki we want to switch the logging agent (ex: to Grafana Agents when it is more mature).
```
```suggestion
I would suggest using the prometheus operator for convinience and then deploying promtail separate to Loki for extra flexibility, so we don't need to touch Loki if we want to switch the logging agent (ex: to Grafana Agents when it is more mature).
```

klausenbusk commented on docs/decisions/0001-grafana-agents.md:
```diff
@@ -0,0 +1,32 @@
+
+Reasons for not going with Grafana agents
+- The documentation is confusing and the Agents and operator docs aren't aligened.
+- Some of the main sellingpoints (integrations and traces) aren't supported by the operator
```
```suggestion
- Some of the main selling points (integrations and traces) aren't supported by the operator
   - We also expected one Grafana Agent per node for scrapping metrics to avoid a single-point-of-failure, but Grafana Agents aren't deployed that way
```

klausenbusk commented on docs/decisions/0001-grafana-agents.md:
```diff
@@ -0,0 +1,32 @@
+- The prometheus operator could also be worth looking at since it give us some of the same features as the agents does, while still providing higher configuration options.
+  (Also endorsment for prometheus-operator is higher - 6,7k stars vs 594 at the time of writing)
+- For Promtail, I am currently seeing issues related to `filetarget.fsnotify.newwatcher: to many open files` which is said to be fixed by using `fs.inotify.max_user_instances` but the Grafana agents does not have a way of adding this to the logging pods, this promtail does [here](https://github.com/grafana/helm-charts/blob/e64169fab5fe248787bd62f4e1a990760fbf99a8/charts/promtail/values.yaml#L20)
+- In the current Agent configuration, we're using it as a package deal for Promtail and Prometheus.
+  - We send all logs to one place
+  - We want all metrics to be scraped by one agent anyways.
```
```suggestion
```

klausenbusk reviewed (approved):
LGTM 🎉

skovbakke commented on docs/decisions/0001-grafana-agents.md:
```diff
@@ -0,0 +1,29 @@
+## Grafana agents - go/no go?
+
+Reasons for not going with Grafana agents
+- The documentation is confusing and the Agents and operator docs aren't aligened.
+- Some of the main selling points (integrations and traces) aren't supported by the operator
+- Generally there were less configuration options in the Grafana Agent Operators.
+  - One example is the implementation of Promtail: The `fs.inotify.max_user_instances` value can be changed in Promtails chart, but is not mentioned in the Grafana Agents implementation.
+  - It is also hard to make changes directly to one of the pods using arguments, you cannot add a argurment, you have to provide **all** the arguments if you want to add or change one.
+
+Grafana Agents was chosen initially as a way to avoid a builky Prometheus instance, as we only need the `remote write` functionality, so we can send metrics to Cortex.
```
```suggestion
## Grafana agents - Are not mature

Reasons for not going with Grafana agents
- The documentation is confusing and the Agents and operator docs are not aligned.
- Some of the main selling points (integrations and traces) are not supported by the operator
- Generally there were less configuration options in the Grafana Agent Operators.
  - One example is the implementation of Promtail: The `fs.inotify.max_user_instances` value can be changed in Promtails chart, but is not mentioned in the Grafana Agents implementation.
  - It is also hard to make changes directly to one of the pods using arguments, you cannot add a argument, you have to provide **all** the arguments if you want to add or change one.

Grafana Agents was chosen initially as a way to avoid a bulky Prometheus instance, as we only need the `remote write` functionality, so we can send metrics to Cortex.
```

skovbakke commented on docs/decisions/0001-grafana-agents.md:
```diff
@@ -0,0 +1,29 @@
+
+### Reasons for switching back to Prometheus and Promtail
+
+- Prometheus has an experimental agent feature, and according to [this blog](https://prometheus.io/blog/2021/11/16/agent/) it has been tested at scale.
+ It sounds like it would achive the same goal of giving us a instance that can `remote write` to cortex, without storing metrics locally.
+- Prometheus is more mature for the time being.
+- The prometheus operator could also be worth looking at since it give us some of the same features as the agents does, while still providing higher configuration options.
+  (Also endorsment for prometheus-operator is higher - 6,7k stars vs 594 at the time of writing)
+- For Promtail, I am currently seeing issues related to `filetarget.fsnotify.newwatcher: to many open files` which is said to be fixed by using `fs.inotify.max_user_instances` but the Grafana agents does not have a way of adding this to the logging pods, this promtail does [here](https://github.com/grafana/helm-charts/blob/e64169fab5fe248787bd62f4e1a990760fbf99a8/charts/promtail/values.yaml#L20)
```
```suggestion
- Prometheus has an experimental agent feature, and according to [this blog](https://prometheus.io/blog/2021/11/16/agent/) it has been tested at scale.
 It sounds like it would achieve the same goal of giving us a instance that can `remote write` to cortex without storing metrics locally.
- Prometheus is more mature for the time being.
- The Prometheus operator could also be worth looking at since it give us some of the same features as the agents does, while still providing more configuration options.
  (Also endorsment for prometheus-operator is higher - 6,7k stars vs 594 at the time of writing)
- For Promtail, there is currently an issue related to `filetarget.fsnotify.newwatcher: to many open files` which is said to be fixed by using `fs.inotify.max_user_instances` but the Grafana agents does not have a way of adding this to the logging pods, this promtail does [here](https://github.com/grafana/helm-charts/blob/e64169fab5fe248787bd62f4e1a990760fbf99a8/charts/promtail/values.yaml#L20)
```

skovbakke commented on docs/decisions/0001-grafana-agents.md:
```diff
@@ -0,0 +1,29 @@
+
+### Recommendation
+
+So from my point of view we should consider going back and use prometheus and promtail,
+
+This is manily due to the inconsistent doc and early stage of the grafana agents, and we need something that is more mature and has more consisten doc.
+
+I would suggest using the prometheus operator for convinience and then deploying promtail separate to Loki for extra flexibility, so we don't need to touch Loki if we want to switch the logging agent (ex: to Grafana Agents when it is more mature).
```
```suggestion
We should consider going back to using Prometheus and Promtail.

This is mainly due to the inconsistent documentation and early stage of the Grafana Agents, and we need something that is mature and has more consistent documentation.

The best option is to use the Prometheus operator for convenience, and then to deploy Promtail separately from Loki for extra flexibility. This way we do not need to change the way we deploy Loki if we want to switch the logging agent at a later stage (e.g. to Grafana Agents when it is more mature).
```

skovbakke reviewed (commented):
A few misspellings, and I suggest we write these documents in a general voice, e.g. not with "I think ..."

MadsHT commented:
> A few misspellings, and I suggest we write these documents in a general voice, e.g. not with "I think ..."

I agree with the general voice and have updated the document.

skovbakke reviewed (approved):
LGTM 🥳

klausenbusk reviewed (approved):
LGTM 🎉
